File: minting/src/main/scala/co/topl/minting/interpreters/Staking.scala
package co.topl.minting.interpreters

import cats.data.OptionT
import cats.effect._
import cats.implicits._
import co.topl.brambl.models.LockAddress
import co.topl.catsutils._
import co.topl.codecs.bytes.tetra.instances._
import co.topl.codecs.bytes.typeclasses.implicits._
import co.topl.consensus.algebras._
import co.topl.consensus.models._
import co.topl.consensus.thresholdEvidence
import co.topl.crypto.hash.Blake2b256
import co.topl.crypto.signing.Ed25519
import co.topl.minting.algebras._
import co.topl.minting.models.VrfHit
import co.topl.models._
import co.topl.typeclasses.implicits._
import com.google.protobuf.ByteString
import org.typelevel.log4cats.Logger
import org.typelevel.log4cats.SelfAwareStructuredLogger
import org.typelevel.log4cats.slf4j.Slf4jLogger

object Staking {

  def make[F[_]: Async](
    a:                        StakingAddress,
    rewardAddress:            LockAddress,
    vkVrf:                    ByteString,
    operationalKeyMaker:      OperationalKeyMakerAlgebra[F],
    consensusState:           ConsensusValidationStateAlgebra[F],
    etaCalculation:           EtaCalculationAlgebra[F],
    ed25519Resource:          Resource[F, Ed25519],
    blake2b256Resource:       Resource[F, Blake2b256],
    vrfCalculator:            VrfCalculatorAlgebra[F],
    leaderElectionValidation: LeaderElectionValidationAlgebra[F],
    protocolVersion:          ProtocolVersion
  ): Resource[F, StakingAlgebra[F]] =
    Resource
      .pure {
        val _rewardAddress = rewardAddress
        new StakingAlgebra[F] {
          implicit private val logger: SelfAwareStructuredLogger[F] =
            Slf4jLogger.getLoggerFromClass[F](Staking.getClass)
          val address: F[StakingAddress] = a.pure[F]
          val rewardAddress: F[LockAddress] = _rewardAddress.pure[F]

          def elect(parentSlotId: SlotId, slot: Slot): F[Option[VrfHit]] = (
            for {
              eta <- OptionT.liftF(etaCalculation.etaToBe(parentSlotId, slot))
              relativeStake <- OptionT(
                consensusState
                  .operatorRelativeStake(parentSlotId.blockId, slot)(a)
                  .warnIfSlow("Local Operator Relative Stake")
              ).flatTapNone(
                Logger[F].debug(s"No stake at slot=$slot")
              )
              threshold <- OptionT.liftF(
                leaderElectionValidation
                  .getThreshold(relativeStake, slot - parentSlotId.slot)
                  .warnIfSlow("Local threshold")
              )
              testProof <- OptionT.liftF(vrfCalculator.proofForSlot(slot, eta))
              rho       <- OptionT.liftF(vrfCalculator.rhoForSlot(slot, eta))
              isLeader <- OptionT.liftF(
                leaderElectionValidation
                  .isSlotLeaderForThreshold(threshold)(rho)
                  .warnIfSlow("Local Operator Is Slot Leader")
              )
              _ <- OptionT.liftF(
                Logger[F].debug(
                  show"Eligibility at" +
                  show" slot=$slot" +
                  show" parentId=${parentSlotId.blockId}" +
                  show" parentSlot=${parentSlotId.slot}" +
                  show" eligible=$isLeader" +
                  show" relativeStake=$relativeStake"
                )
              )
              vrfHit <- OptionT
                .whenF[F, VrfHit](isLeader)(
                  blake2b256Resource
                    .use(implicit b => Sync[F].delay(thresholdEvidence(threshold)))
                    .map(evidence =>
                      VrfHit(
                        EligibilityCertificate(testProof, vkVrf, evidence, eta.data),
                        slot,
                        threshold
                      )
                    )
                )
            } yield vrfHit
          ).value

          def certifyBlock(
            parentSlotId:         SlotId,
            slot:                 Slot,
            unsignedBlockBuilder: UnsignedBlockHeader.PartialOperationalCertificate => UnsignedBlockHeader,
            eta:                  Eta
          ): F[Option[BlockHeader]] =
            OptionT(operationalKeyMaker.operationalKeyForSlot(slot, parentSlotId, eta)).semiflatMap {
              operationalKeyOut =>
                for {
                  partialCertificate <- Sync[F].delay(
                    UnsignedBlockHeader.PartialOperationalCertificate(
                      operationalKeyOut.parentVK,
                      operationalKeyOut.parentSignature,
                      operationalKeyOut.childVK
                    )
                  )
                  unsignedBlock = unsignedBlockBuilder(partialCertificate)
                  messageToSign = unsignedBlock.signableBytes.toByteArray
                  signature <- ed25519Resource.use(ed25519 =>
                    Sync[F].delay(
                      ed25519.sign(
                        Ed25519.SecretKey(operationalKeyOut.childSK.toByteArray),
                        messageToSign
                      )
                    )
                  )
                  operationalCertificate = OperationalCertificate(
                    operationalKeyOut.parentVK,
                    operationalKeyOut.parentSignature,
                    partialCertificate.childVK,
                    ByteString.copyFrom(signature)
                  )
                  header = BlockHeader(
                    headerId = None,
                    unsignedBlock.parentHeaderId,
                    unsignedBlock.parentSlot,
                    unsignedBlock.txRoot,
                    unsignedBlock.bloomFilter,
                    unsignedBlock.timestamp,
                    unsignedBlock.height,
                    unsignedBlock.slot,
                    unsignedBlock.eligibilityCertificate,
                    operationalCertificate,
                    unsignedBlock.metadata,
                    unsignedBlock.address,
                    protocolVersion
                  )
                } yield header
            }.value
        }
      }
}

File: dart/minting/lib/interpreters/staking.dart
import 'package:bifrost_codecs/codecs.dart';
import 'package:bifrost_common/models/unsigned.dart';
import 'package:bifrost_consensus/algebras/consensus_validation_state_algebra.dart';
import 'package:bifrost_consensus/algebras/eta_calculation_algebra.dart';
import 'package:bifrost_consensus/algebras/leader_election_validation_algebra.dart';
import 'package:bifrost_consensus/utils.dart';
import 'package:bifrost_crypto/ed25519.dart' as cryptoEd25519;
import 'package:bifrost_minting/algebras/operational_key_maker_algebra.dart';
import 'package:bifrost_minting/algebras/staking_algebra.dart';
import 'package:bifrost_minting/algebras/vrf_calculator_algebra.dart';
import 'package:bifrost_minting/models/vrf_hit.dart';
import 'package:fixnum/fixnum.dart';
import 'package:logging/logging.dart';
import 'package:topl_protobuf/consensus/models/block_header.pb.dart';
import 'package:topl_protobuf/consensus/models/eligibility_certificate.pb.dart';
import 'package:topl_protobuf/consensus/models/operational_certificate.pb.dart';
import 'package:topl_protobuf/consensus/models/slot_data.pb.dart';
import 'package:topl_protobuf/consensus/models/staking_address.pb.dart';

class Staking extends StakingAlgebra {
  final StakingAddress _address;
  final List<int> vkVrf;
  final OperationalKeyMakerAlgebra operationalKeyMaker;
  final ConsensusValidationStateAlgebra consensusValidationState;
  final EtaCalculationAlgebra etaCalculation;
  final VrfCalculatorAlgebra vrfCalculator;
  final LeaderElectionValidationAlgebra leaderElectionValidation;

  final log = Logger("staking");

  Staking(
      this._address,
      this.vkVrf,
      this.operationalKeyMaker,
      this.consensusValidationState,
      this.etaCalculation,
      this.vrfCalculator,
      this.leaderElectionValidation);

  @override
  StakingAddress get address => _address;

  @override
  Future<BlockHeader?> certifyBlock(
      SlotId parentSlotId,
      Int64 slot,
      UnsignedBlockHeader Function(PartialOperationalCertificate)
          unsignedBlockBuilder) async {
    final operationalKeyOutOpt =
        await operationalKeyMaker.operationalKeyForSlot(slot, parentSlotId);
    if (operationalKeyOutOpt != null) {
      final partialCertificate = PartialOperationalCertificate(
        operationalKeyOutOpt.parentVK,
        operationalKeyOutOpt.parentSignature,
        operationalKeyOutOpt.childKeyPair.vk,
      );
      final unsignedHeader = unsignedBlockBuilder(partialCertificate);
      final List<int> messageToSign = unsignedHeader.signableBytes;
      final cryptoKeyPair = await cryptoEd25519.Ed25519KeyPair(
          operationalKeyOutOpt.childKeyPair.sk,
          operationalKeyOutOpt.childKeyPair.vk);
      final operationalCertificate = OperationalCertificate(
        parentVK: operationalKeyOutOpt.parentVK,
        parentSignature: operationalKeyOutOpt.parentSignature,
        childVK: operationalKeyOutOpt.childKeyPair.vk,
        childSignature: await cryptoEd25519.ed25519.signKeyPair(
          messageToSign,
          cryptoKeyPair,
        ),
      );
      final header = BlockHeader(
        parentHeaderId: unsignedHeader.parentHeaderId,
        parentSlot: unsignedHeader.parentSlot,
        txRoot: unsignedHeader.txRoot,
        bloomFilter: unsignedHeader.bloomFilter,
        timestamp: unsignedHeader.timestamp,
        height: unsignedHeader.height,
        slot: unsignedHeader.slot,
        eligibilityCertificate: unsignedHeader.eligibilityCertificate,
        operationalCertificate: operationalCertificate,
        metadata: unsignedHeader.metadata,
        address: unsignedHeader.address,
      );
      return header;
    }
    return null;
  }

  @override
  Future<VrfHit?> elect(SlotId parentSlotId, Int64 slot) async {
    final eta = await etaCalculation.etaToBe(parentSlotId, slot);
    final relativeStake = await consensusValidationState.operatorRelativeStake(
        parentSlotId.blockId, slot, address);
    if (relativeStake == null) return null;
    final threshold = await leaderElectionValidation.getThreshold(
        relativeStake, slot - parentSlotId.slot);
    final rho = await vrfCalculator.rhoForSlot(slot, eta);
    final isLeader =
        await leaderElectionValidation.isSlotLeaderForThreshold(threshold, rho);
    log.fine("Staking leader election test result=$isLeader slot=$slot");
    if (isLeader) {
      final evidence = threshold.thresholdEvidence;
      final testProof = await vrfCalculator.proofForSlot(slot, eta);
      final cert = EligibilityCertificate(
          vrfSig: testProof,
          vrfVK: vkVrf,
          thresholdEvidence: evidence,
          eta: eta);
      return VrfHit(cert, slot, threshold);
    }
    return null;
  }
}
